File: cpp/src/two_sum.cpp
#include "../inc/solution.h"
#include <unordered_map>
#include <vector>

std::vector<int> Solution::twoSum(std::vector<int>& nums, int target)
{
   std::unordered_map<int, int> map;
   for (int i = 0; i < nums.size(); ++i)
   {
      int complement = target - nums[i];
      if (map.count(complement))
      {
         return {map[complement], i};
      }
      map[nums[i]] = i;
   }
   return {};
}

#include <gtest/gtest.h>

TEST(TwoSumTest, Example1)
{
   std::vector<int> nums = { 2, 7, 11, 15 };
   int target = 9;
   auto result = Solution().twoSum(nums, target);
   EXPECT_EQ(result, (std::vector<int>{ 0, 1 }));
}

TEST(TwoSumTest, Example2)
{
   std::vector<int> nums = { 3, 2, 4 };
   int target = 6;
   auto result = Solution().twoSum(nums, target);
   EXPECT_EQ(result, (std::vector<int>{ 1, 2 }));
}

TEST(TwoSumTest, Example3)
{
   std::vector<int> nums = { 3, 3 };
   int target = 6;
   auto result = Solution().twoSum(nums, target);
   EXPECT_EQ(result, (std::vector<int>{ 0, 1 }));
}

File: rust/src/problems/easy/two_sum.rs
use crate::Solution;
use std::collections::HashMap;
use std::vec;

impl Solution {
   pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
      let mut map: HashMap<i32, i32> = HashMap::new();
      for (i, num) in nums.iter().enumerate() {
         let diff = target - num;
         if map.contains_key(&diff) {
            return vec![map[&diff], i as i32]
         }
         map.insert(*num, i as i32);
      }
      vec![]
   }
}

#[cfg(test)]
mod tests {
   use crate::Solution;

   #[test]
   fn example_1() {
      assert_eq!(Solution::two_sum(vec![2, 7, 11, 15], 9), vec![0, 1]);
   }

   #[test]
   fn example_2() {
      assert_eq!(Solution::two_sum(vec![3, 2, 4], 6), vec![1, 2]);
   }

   #[test]
   fn example_3() {
      assert_eq!(Solution::two_sum(vec![3, 3], 6), vec![0, 1]);
   }
}